### DISCUSSION TITLE
Security Testing - Ikev2 cipher suite and other options enumeration

### DISCUSSION THREAD

**Discussion by Himself132** [strongswan/strongswan#442] [General]
Hello,

I am a long time member of the security/ethical hacking community and have a lot of experience evaluating company/org vpn implementations and other protocols and related cipher suite choices and strengths which is, as you know one of the most important aspects of data confidentiality and integrity and availability.  When evaluating IPSec a lot of the analysis that goes into it from a security perspective is enumerating options in cipher suites for protocols like ikev1 and ikev2.  ike-scan is a tool that many many security professionals rely on for IPSec evaluation.  Seeing strongswan and setting up a configuration file and starting up the services with various cipher suite settings and other options one can individually enumerate those values and I was trying to think of an easier way to do it, with a script that starts and stops the service and something to script to create files for the starting and stopping services to run and output to the logs.  Then I got to thinking that maybe strongswan has some test scripts that would accomplish the same thing for QA/new releases.  Do you have any test scripts that'll run against a target utilizing ikev2 where you run through all the cipher suites and options that the security community could leverage or some starting point that I could use and contribute to the security community by placing it on github and maintaining it myself somehow?

Any help would be greatly appreciated.

**Comment by tobiasbrunner** [comment]
I'm not aware of any existing scripts like that. But I guess you could use [vici](https://wiki.strongswan.org/projects/strongswan/wiki/Vici) to write something like that (it's definitely easier than creating config files etc.).

**Reply by Himself132** [reply]
Thank you Tobias, i will check it out, i did some additional investigation and saw that someone wrote some code down for ikev2 in scapy also, should be able to figure something out.  Although i am a bit surprised you do not have these scripts for testing purposes of your own.  If i do end up throwing something together would you be interested?  I like to try and contribute back to projects that help me out - of course no promises on my end. https://scapy.readthedocs.io/en/latest/api/scapy.contrib.ikev2.html

**Reply by tobiasbrunner** [reply]
> Although i am a bit surprised you do not have these scripts for testing purposes of your own.

What purpose are you referring to exactly?

**Comment by Himself132** [comment]
What I mean is, when you make upgrades or check for interoperability with other VPN endpoint technologies, it'd seem to make sense to test new versions or releases against them to ensure there are no issues as part of QA.  I definitely don't live in the dev world so excuse me if that it pie-in-the-sky mentality or unrealistic or doesn't make sense. I'm just in the security world trying to leverage work already done by others :) and appreciate your talking through it!

**Reply by tobiasbrunner** [reply]
> What I mean is, when you make upgrades

There are unit tests that e.g. check cryptographic implementations against test vectors and we have [regression tests](https://www.strongswan.org/testing/testresults/) that also test some of them in practice.

> or check for interoperability with other VPN endpoint technologies

We only very rarely do that (mainly during scheduled interoperability workshops for new protocol extensions).

But a script like you described (that tests algorithm combinations in proposals) seems unrelated to either as it's mostly configuration and not implementation dependent. So I don't really see a value in such a script for us as developers. Some users might find it useful if they don't know anything about the remote peer. However, IKEv2 made proposal negotiation a lot easier than IKEv1 as it allows sending multiple algorithms per transform type in a proposal, so it's not necessary to match the peer's proposal exactly (our default proposals, one for regular and one for AEAD algorithms, basically contain all supported algorithms that are not deemed too weak, so if the peer supports at least one of each type, the connection gets established).

**Reply by Himself132** [reply]
thank you, makes perfect sense ,thanks for the clarification
